src/System.Drawing.Common/src/System/Drawing/LibraryResolver.cs

```diff
@@ -0,0 +1,32 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Runtime.InteropServices;
+
+namespace System.Drawing.Common.System.Drawing
+{
+    internal static class LibraryResolver
+    {
+        internal static Dictionary<string, Func<IntPtr>> LibraryLoaders = new Dictionary<string, Func<IntPtr>>();
+
+        static LibraryResolver()
+        {
+            // Hook our custom resolver
+            NativeLibrary.SetDllImportResolver(Assembly.GetExecutingAssembly(), DllImportResolver);
+        }
+
+        private static IntPtr DllImportResolver(string libraryName, Assembly assembly, DllImportSearchPath? searchPath)
+        {
+            if (LibraryLoaders.TryGetValue(libraryName, out Func<IntPtr> loader))
```

Comment:
I do not see anything that guarantees thread safety here. I do not think this needs to be a fancy registration scheme. Just check for the two libraries we care about and call the appropriate method

Reply:
How do you suggest I ensure we are hooked dynamically when the relevant wrapping classes are accessed? I was relying on the dictionary access from the static constructors from the two classes. I'm also unclear on how threading would bust as it currently is set up.

Reply:
My presumption is that the callback results are cached for positive results. Is that true @jkoritzinsky?

Reply:
> I'm also unclear on how threading would bust as it currently is set up.

The two static constructors that register the LibraryLoaders can run in parallel on different threads; or they can run in parallel while the other thread is resolving the library.

```
class LibraryResolver
{
     static LibraryResolver()
     {
         NativeLibrary.SetDllImportResolver(typeof(LibraryResolver).Assembly, DllImportResolver);
     }

      private static IntPtr DllImportResolver(string libraryName, Assembly assembly, DllImportSearchPath? searchPath)
      {
           if (LibcupsNative.LibraryName == "libcups") return LibcupsNative.LoadLibcups();
           if (libraryName == "...") return ...
           return IntPtr.Zero;
      }

     void EnsureRegistered()
     {
          // dummy call to trigger the static constructor
     }
}
```

Then call `LibraryResolver.EnsureRegistered();` in the static constructor of each of `*Native` types that depend on it.